### PR TITLE
Sensor configuration and management overhaul

### LINKED ISSUE #15: Improve Raspberry Pi support
At the moment, there are a few elements which need to be worked on to ensure it's fully compatible with the Raspberry Pi (3/4):

- [x] Ensure installer works on Raspberry Pi and properly sets up SIGHTS
- [x] Ensure I2C sensors work (including multiple sensors at the same time) on the Raspberry Pi and add any additional steps required to enable I2C to the installer.
- [x] Make the installer install ARM binaries for Motion on the Raspberry Pi
- [x] Test to ensure all software components work as expected.

### LINKED ISSUE #36: Overhaul sensor configuration and management
**Is your feature request related to a problem? Please describe.**
At the moment, the `temperature` key in the config file (within the `sensors` section) will use the MLX90614 wrapper. I've written untested support for a different temp sensor as an example in some new documentation I'm writing and we could even make it a supported sensor. The issue is that `temperature` is hardcoded to mlx90614.

**Describe the solution you'd like**
We could change it to `mlx90614` of course, that would fix the problem. This would mean using the sensor name as the key in the JSON file for sensors. So, `bme280` for the new type of temperature sensor and `mlx90614` for the type we use.

However, this still doesn't allow for multiple sensors of the same type.

My proposed solution would be to change the `sensors` section of the config file to an array.

Here's an example of how the `sensors` section will look. Note how we have multiple of one type of sensor, as well as multiple different types of temperature sensors.
```json
"sensors": [
	{
		"type": "amg8833",
		"name": "Thermal Camera",
		"enabled": true,
		"frequency": 0.5,
		"width": 32,
		"height": 24
	},
	{
		"type": "mlx90614",
		"name": "Temperature Sensor (front)",
		"enabled": true,
		"frequency": 2,
		"address": "0x5A"
	},
	{
		
		"type": "bme280",
		"name": "Temperature Sensor (back)",
		"enabled": true,
		"frequency": 4
	},
	{ 
		"type": "vl53l0x",
		"name": "Distance Sensor (1)",
		"enabled": true,
		"frequency": 2,
		"address": "0x2A"
	},
	{ 
		"type": "vl53l0x",
		"name": "Distance Sensor (2)",
		"enabled": true,
		"frequency": 2,
		"address": "0x2B"
	},
	{
		"type": "sgp30",
		"name": "Gas Sensor",
		"enabled": true,
		"frequency": 2
	}
	{
		"type": "memory",
		"name": "System Memory",
		"enabled": true,
		"frequency": 3
	},
	{
		"type": "cpu_temp",
		"name": "CPU Temperature",
		"enabled": true,
		"frequency": 5
	},
},
```

Additionally, this means we can implement proper sensor management.

`sensor_stream.py` needs to look through the config file, find active sensors, lookup their respective SensorWrapper class (we need to make a basic lookup table) and create a new instance of that class and append it to the `sensors` array

The lookup table basically needs to function like this, that takes the `type` key from the configuration file and returns the appropriate class:

```json
{
	"mlx90614": MLX90614Wrapper,
	"sgp30": SGP30Wrapper,
	"bme280": BME280Wrapper
}
```

### LINKED ISSUE #37: Implement plugin system for sensor wrappers
**Is your feature request related to a problem? Please describe.**
At the moment, adding a new sensor wrapper involves adding it to the `__init__.py` file within the `sensors` directory. 

It's messy, it's not intuitive and making the init file import everything in the directory instead is apparently a [very bad idea](https://stackoverflow.com/questions/16788380/import-all-classes-or-functions-in-all-files-in-folder-as-if-they-were-all-in). In the wise words of one StackOverflow user:
> This is a very bad idea.

**Describe the solution you'd like**
Make `sensor_stream.py` scan the sensors directory and import each file, or plugin, automatically.

I would also propose referring to these files as plugins from hereon, to better clarify their modular nature.

From [this StackOverflow post](https://stackoverflow.com/questions/16788380/import-all-classes-or-functions-in-all-files-in-folder-as-if-they-were-all-in):

> The way you normally do this is to iterate one or more plugin directories with `os.listdir` or `os.walk`, then, for each Python file, use `importlib` (or lower-level equivalents from imp, in Python 2.x) to import it by path.

**Additional context**

Useful reference:

- https://stackoverflow.com/questions/932069/building-a-minimal-plugin-architecture-in-python
- https://lkubuntu.wordpress.com/2012/10/02/writing-a-python-plugin-api/

### PR DESCRIPTION
**Pull request by Conr86**
# Description

Firstly, this implements a plugin system for sensors, so they are dynamically loaded. This means we no longer have to add every wrapper to the main code, and the sensors are loaded from the configuration file dynamically.

Secondly, the configuration file format has been rewritten. It now supports multiple sensors of the same type, and some additional options.

```json
"sensors": [
	{
		"type": "amg8833",
		"name": "Thermal Camera",
		"enabled": false,
		"frequency": 0.5,
		"width": 32,
		"height": 24
	},
	{
		"type": "mlx90614",
		"name": "Temperature Sensor",
		"enabled": false,
		"frequency": 2,
		"address": "0x5A"
	},
	{
		"type": "sgp30",
		"name": "Gas Sensor",
		"enabled": false,
		"frequency": 2
	},
	{
		"type": "memory",
		"name": "System Memory",
		"enabled": true,
		"frequency": 2
	},
	{
		"type": "cpu_temp",
		"name": "CPU Temperature",
		"enabled": true,
		"frequency": 5
	}
],
```

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Re-write of an existing feature
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# Resolves issues

- Closes #36 (sensor config / management overhaul)
- Closes #37 (plugin system)
- Closes #15 (Raspberry Pi support)

# Testing

Everything seems to work just as well as it did beforehand. It's also just as well-tested as it was beforehand.

Which is to say, sparsely. Most of it still hasn't been tested with actual sensor hardware, with the exception of the MLX90614 sensors. Don't worry though, testing the other sensors is high on the to-do list.

This has been tested on a Raspberry Pi, with multiple MLX90614 sensors, and works fine.

This will break the interface side code, until that's completed too. As such I've marked it as a breaking change.

# Final checklist:

- [x] Code follows the style guidelines of SIGHTS
- [x] I have performed a self-review of my own code
- [x] Comments have been added, particularly in hard-to-understand areas
- [x] Made corresponding changes to the documentation
- [x] The changes cause no new errors or warnings

### CODE REVIEW

File: src/requirements.txt [hunk 1]
```diff
@@ -4,5 +4,6 @@ websockets>=7.0
 pyserial>=3.4
 PySGP30
 PyMLX90614
+Adafruit_AMG88xx
```
Comment (WilliamsJack) [inline src/requirements.txt]: I was getting `ModuleNotFoundError: No module named 'Adafruit_AMG88xx'` and `ModuleNotFoundError: No module named 'Adafruit_bitfield'` on this branch after re-running `pip install` on this `requirements.txt` file. I had to use the old AMG88xx library:
`sudo python3 -m pip install -e git+https://github.com/adafruit/Adafruit_AMG88xx_python#egg=Adafruit_AMG88xx`

And this additional library:
`sudo python3 -m pip install -e git+https://github.com/adafruit/Adafruit_bitfield_python#egg=Adafruit_bitfield`

Comment (WilliamsJack) [reply src/requirements.txt]: Interesting, a second attempt with `Adafruit_AMG88xx` worked...

Comment (WilliamsJack) [reply src/requirements.txt]: Works with ``requirements.txt` too now. Going to hide this conversation as if it never happened.

Comment (Conr86) [reply src/requirements.txt]: What a fascinating roller coaster of events. Do you know why it happened?

Comment (WilliamsJack) [reply src/requirements.txt]: Nope, not a clue. As far as I can remember, I uninstalled the old Adafruit library before testing again. If I didn't, or it was unsuccessful, that could be why.

Comment (WilliamsJack) [reply src/requirements.txt]: New environment, encountered this issue again. Installed requirements.txt on this branch (including Adafruit_AMG88xx), service won't start:

```pytb
Traceback (most recent call last):
  File "src/main.py", line 89, in <module>
    manager.run()
  File "src/main.py", line 40, in run
    self.sensor_process = SensorStream(1, self.sensor_pipe, self.config_file)
  File "/opt/sights/SIGHTSRobot/src/sensor_stream.py", line 53, in __init__
    plugin = importlib.import_module(plugin_name)
  File "/usr/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/opt/sights/SIGHTSRobot/src/sensors/amg8833_wrapper.py", line 2, in <module>
    import Adafruit_AMG88xx.Adafruit_AMG88xx as AMG88xx
ModuleNotFoundError: No module named 'Adafruit_AMG88xx'
```

After installing the old libraries mentioned in the original comment, the service was able to start.

`sudo python3 -m pip install -e git+https://github.com/adafruit/Adafruit_AMG88xx_python#egg=Adafruit_AMG88xx`
`sudo python3 -m pip install -e git+https://github.com/adafruit/Adafruit_bitfield_python#egg=Adafruit_bitfield`

Comment (WilliamsJack) [reply src/requirements.txt]: Resolved by #40 for new installations (or existing ones after running `pip install` against `requirements.txt` again) should you experience the same issue.

**Pull request merged**